### PR TITLE
Add Route Agent health check

### PR DESCRIPTION
**Pull request by aswinsuryan**
Add Route Agent health check

### CODE REVIEW

File: pkg/spoke/submarineragent/connections_controller.go [hunk 1]
```diff
@@ -79,20 +85,70 @@ func (c *connectionsStatusController) sync(ctx context.Context, syncCtx factory.
...
+	updatedStatus, updated, err = addon.UpdateStatus(ctx, c.addOnClient, c.clusterName, addon.UpdateConditionFn(finalCondition))
+	if err != nil {
+		return err
+	}
+
+	if updated {
+		c.logger.Infof("Updated final RouteAgent ManagedClusterAddOn status condition: %s", resource.ToJSON(finalCondition))
+		syncCtx.Recorder().Eventf("ManagedClusterAddOnStatusUpdated", "Updated RouteAgent condition: %#v",
+			updatedStatus.Conditions)
+	}
```
Comment (tpantelis) [inline pkg/spoke/submarineragent/connections_controller.go:150]: We can combine the two conditions in one `UpdateStatus` call:
```suggestion
	updatedStatus, updated, err = addon.UpdateStatus(ctx, c.addOnClient, c.clusterName, addon.UpdateConditionFn(gatewaycondition),
	addon.UpdateConditionFn(routeAgentCondition))
	if err != nil {
		return err
	}

	if updated {
		c.logger.Infof("Updated submariner ManagedClusterAddOn status condition: %s", resource.ToJSON(updatedStatus.Conditions))
		syncCtx.Recorder().Eventf("ManagedClusterAddOnStatusUpdated", "Updated status condition: %#v",
			updatedStatus.Conditions)
	}
```

File: pkg/spoke/submarineragent/connections_controller.go [hunk 2]
```diff
@@ -153,3 +209,46 @@ func (c *connectionsStatusController) checkSubmarinerConnections(submariner *sub
...
+
+	var remoteEndpoints []submarinermv1.RemoteEndpoint
+	if len(routeAgent.Status.RemoteEndpoints) > 0 {
+		remoteEndpoints = routeAgent.Status.RemoteEndpoints
+	}
+
+	var unhealthyMessages []string
+
+	for i := range remoteEndpoints {
```
Comment (tpantelis) [inline pkg/spoke/submarineragent/connections_controller.go:225]: We can simplify:

```suggestion
	var unhealthyMessages []string

        remoteEndpoints := routeAgent.Status.RemoteEndpoints
	for i := range remoteEndpoints {
```

File: pkg/spoke/submarineragent/connections_controller_test.go [hunk 3]
```diff
@@ -120,13 +125,102 @@ var _ = Describe("Connections Status Controller", func() {
 			})
 		})
 	})
+
+	When("when all RouteAgents have healthy connections", func() {
+		It("should set the  condition to ConnectionsHealthy", func() {
+			t.awaitConnectionsEstablishedStatusCondition()
+		})
```
Comment (tpantelis) [inline pkg/spoke/submarineragent/connections_controller_test.go:132]:
```
func (t *connStatusControllerTestDriver) awaitRouteAgentsEstablishedStatusCondition() {
	t.awaitRouteAgentStatusCondition(metav1.ConditionTrue, "ConnectionsEstablished")
}
```

```suggestion
		It("should update the ManagedClusterAddOn status condition to connections established", func() {
			t.awaitRouteAgentsEstablishedStatusCondition()
		})
```

File: pkg/spoke/submarineragent/connections_controller_test.go [hunk 4]
```diff
@@ -120,13 +125,102 @@ var _ = Describe("Connections Status Controller", func() {
...
+								StatusMessage: "Attempting connection",
+							},
+						},
+					},
+				},
+			}
+			t.updateRouteAgent(t.routeAgent[0])
+		})
+
+		It("should report the ManagedClusterAddOn status as progressing", func() {
```
Comment (tpantelis) [inline pkg/spoke/submarineragent/connections_controller_test.go:157]:
```suggestion
		It("should update the ManagedClusterAddOn status condition to degraded", func() {
```

File: pkg/spoke/submarineragent/connections_controller_test.go [hunk 5]
```diff
@@ -120,13 +125,102 @@ var _ = Describe("Connections Status Controller", func() {
...
+							{
+								Status:        submv1.Connecting,
+								Spec:          submv1.EndpointSpec{Hostname: "remote1"},
+								StatusMessage: "Attempting connection",
+							},
+						},
+					},
+				},
+			}
+			t.updateRouteAgent(t.routeAgent[0])
```
Comment (tpantelis) [inline pkg/spoke/submarineragent/connections_controller_test.go:154]:
```suggestion
			t.routeAgent[0].Status.RemoteEndpoints[0].Status = submv1.Connecting
```

File: pkg/spoke/submarineragent/connections_controller_test.go [hunk 6]
```diff
@@ -120,13 +125,102 @@ var _ = Describe("Connections Status Controller", func() {
...
+		})
+
+		It("should set the  condition to ConnectionsHealthy", func() {
+			t.awaitConnectionsEstablishedStatusCondition()
+		})
```
Comment (tpantelis) [inline pkg/spoke/submarineragent/connections_controller_test.go:186]:
```suggestion
		BeforeEach(func() {
			t.routeAgent[0].Status.RemoteEndpoints[0].Status = submv1.ConnectionNone
		})

		It("should update the ManagedClusterAddOn status condition to connections established", func() {
			t.awaitRouteAgentStatusCondition(metav1.ConditionTrue, "ConnectionsEstablished")
		})
```

File: pkg/spoke/submarineragent/connections_controller_test.go [hunk 7]
```diff
@@ -120,13 +125,102 @@ var _ = Describe("Connections Status Controller", func() {
...
+						},
+					},
+				},
+			}
+			t.updateRouteAgent(t.routeAgent[0])
+		})
+
+		It("should set the  condition to ConnectionsHealthy", func() {
+			t.awaitRouteAgentsDegradedStatusCondition()
+		})
```
Comment (tpantelis) [inline pkg/spoke/submarineragent/connections_controller_test.go:213]:
```suggestion
		BeforeEach(func() {
			t.routeAgent[0].Status.RemoteEndpoints[0].Status = submv1.ConnectionError
		})

		It("should update the ManagedClusterAddOn status condition to degraded", func() {
			t.awaitRouteAgentsDegradedStatusCondition()
		})
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/spoke/submarineragent/connections_controller.go [hunk 8]
```diff
@@ -23,27 +25,31 @@ import (
...
+	submarinerLister   cache.GenericLister
+	spokeDynamicClient dynamic.Interface
+	clusterName        string
+	logger             log.Logger
 }
 
 // NewConnectionsStatusController returns an instance of submarinerAgentStatusController.
 func NewConnectionsStatusController(clusterName string, addOnClient addonclient.Interface, submarinerInformer informers.GenericInformer,
-	recorder events.Recorder,
+	spokeDynamicClient dynamic.Interface, recorder events.Recorder,
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: Done.

File: pkg/spoke/submarineragent/connections_controller.go [hunk 9]
```diff
@@ -79,20 +85,70 @@ func (c *connectionsStatusController) sync(ctx context.Context, syncCtx factory.
...
 
 		syncCtx.Recorder().Eventf("ManagedClusterAddOnStatusUpdated", "Updated status conditions:  %#v",
 			updatedStatus.Conditions)
 	}
 
+	routeAgentGVR := schema.GroupVersionResource{
+		Group:    "submariner.io",
+		Version:  "v1",
+		Resource: "routeagents",
+	}
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: not required now

File: pkg/spoke/submarineragent/connections_controller_test.go [hunk 10]
```diff
@@ -171,6 +265,72 @@ func newConnStatusControllerTestDriver() *connStatusControllerTestDriver {
...
+			unstructuredRouteAgent.SetGroupVersionKind(routeAgentGVR.GroupVersion().WithKind("RouteAgent"))
+
+			// Create the resource using spokeDynamicClient
+			_, err = t.spokeDynamicClient.Resource(routeAgentGVR).Namespace(t.routeAgent[i].Namespace).Create(
+				context.TODO(),
+				unstructuredRouteAgent,
+				metav1.CreateOptions{},
+			)
+			Expect(err).To(Succeed())
+		}
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller_test.go]: Done.

File: pkg/spoke/submarineragent/connections_controller.go [hunk 11]
```diff
@@ -79,20 +85,70 @@ func (c *connectionsStatusController) sync(ctx context.Context, syncCtx factory.
...
+		routeAgentCondition := c.checkRouteAgentConditions(routeAgent)
+		if routeAgentCondition.Status == metav1.ConditionTrue && routeAgentCondition.Reason == connectionsDegraded {
+			allUnhealthyMessages = append(allUnhealthyMessages, routeAgentCondition.Message)
+		}
+	}
+
+	finalCondition := &metav1.Condition{
+		Type:   routeAgentConnectionDegraded,
+		Status: metav1.ConditionFalse,
+		Reason: "AllConnectionsHealthy",
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: Done.

File: pkg/spoke/submarineragent/connections_controller.go [hunk 12]
```diff
@@ -79,20 +85,70 @@ func (c *connectionsStatusController) sync(ctx context.Context, syncCtx factory.
...
+	for _, item := range routeAgentList.Items {
+		routeAgent := &submarinermv1.RouteAgent{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &routeAgent); err != nil {
+			return err
+		}
+
+		routeAgentCondition := c.checkRouteAgentConditions(routeAgent)
+		if routeAgentCondition.Status == metav1.ConditionTrue && routeAgentCondition.Reason == connectionsDegraded {
+			allUnhealthyMessages = append(allUnhealthyMessages, routeAgentCondition.Message)
+		}
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: Done.

File: pkg/spoke/submarineragent/connections_controller.go [hunk 13]
```diff
@@ -79,20 +85,70 @@ func (c *connectionsStatusController) sync(ctx context.Context, syncCtx factory.
...
+	routeAgentList, err := c.spokeDynamicClient.Resource(routeAgentGVR).Namespace(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	var allUnhealthyMessages []string
+
+	for _, item := range routeAgentList.Items {
+		routeAgent := &submarinermv1.RouteAgent{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &routeAgent); err != nil {
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: Done.

File: pkg/spoke/submarineragent/connections_controller.go [hunk 14]
```diff
@@ -79,20 +85,70 @@ func (c *connectionsStatusController) sync(ctx context.Context, syncCtx factory.
...
+		Status: metav1.ConditionFalse,
+		Reason: "AllConnectionsHealthy",
+	}
+
+	if len(allUnhealthyMessages) > 0 {
+		finalCondition.Status = metav1.ConditionTrue
+		finalCondition.Reason = connectionsDegraded
+		finalCondition.Message = strings.Join(allUnhealthyMessages, "\n")
+	} else {
+		finalCondition.Message = "All RouteAgent connections to remote endpoints are established and healthy."
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: Done.

File: pkg/spoke/submarineragent/connections_controller.go [hunk 15]
```diff
@@ -79,20 +85,70 @@ func (c *connectionsStatusController) sync(ctx context.Context, syncCtx factory.
...
+			return err
+		}
+
+		routeAgentCondition := c.checkRouteAgentConditions(routeAgent)
+		if routeAgentCondition.Status == metav1.ConditionTrue && routeAgentCondition.Reason == connectionsDegraded {
+			allUnhealthyMessages = append(allUnhealthyMessages, routeAgentCondition.Message)
+		}
+	}
+
+	finalCondition := &metav1.Condition{
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: Done.

File: pkg/spoke/submarineragent/connections_controller.go (hunk 1 above)
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: Done.

File: pkg/spoke/submarineragent/connections_controller_test.go (hunk 3 above)
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller_test.go]: Done.

File: pkg/spoke/submarineragent/connections_controller_test.go (hunk 4 above)
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller_test.go]: Done.

File: pkg/spoke/submarineragent/connections_controller_test.go (hunk 5 above)
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller_test.go]: Done.

File: pkg/spoke/submarineragent/connections_controller_test.go (hunk 6 above)
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller_test.go]: Done.

File: pkg/spoke/submarineragent/connections_controller_test.go (hunk 7 above)
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller_test.go]: Done.

File: pkg/spoke/submarineragent/connections_controller.go (hunk 2 above)
Comment (aswinsuryan) [reply pkg/spoke/submarineragent/connections_controller.go]: Done.

**Review by tpantelis — APPROVED** [review]

**Pull request merged**